pkg/controller/job/job_controller.go

```diff
@@ -0,0 +1,452 @@
+		jm.expectations.DeleteExpectations(key)
+		return nil
+	}
+	if err != nil {
+		glog.Infof("Unable to retrieve job %v from store: %v", key, err)
+		jm.queue.Add(key)
+		return err
+	}
+	job := *obj.(*experimental.Job)
+	if !jm.podStoreSynced() {
+		// Sleep so we give the pod reflector goroutine a chance to run.
+		time.Sleep(replicationcontroller.PodStoreSyncedPollPeriod)
+		glog.Infof("Waiting for pods controller to sync, requeuing job %v", job.Name)
+		jm.enqueueController(&job)
+		return nil
+	}
+
+	// Check the expectations of the job before counting active pods, otherwise a new pod can sneak in
+	// and update the expectations after we've retrieved active pods from the store. If a new pod enters
+	// the store after we've checked the expectation, the job sync is just deferred till the next relist.
+	jobKey, err := controller.KeyFunc(&job)
+	if err != nil {
+		glog.Errorf("Couldn't get key for job %#v: %v", job, err)
+		return err
+	}
+	jobNeedsSync := jm.expectations.SatisfiedExpectations(jobKey)
+	podList, err := jm.podStore.Pods(job.Namespace).List(labels.Set(job.Spec.Selector).AsSelector())
+	if err != nil {
+		glog.Errorf("Error getting pods for job %q: %v", key, err)
+		jm.queue.Add(key)
+		return err
+	}
+
+	activePods := controller.FilterActivePods(podList.Items)
+	active := len(activePods)
+	successful, unsuccessful := getStatus(jobKey, job.Spec.Template.Spec.RestartPolicy, podList.Items)
+	if jobNeedsSync {
+		active = jm.manageJob(activePods, successful, unsuccessful, &job)
+	}
+	completions := successful
+	if job.Spec.Template.Spec.RestartPolicy == api.RestartPolicyNever {
+		completions += unsuccessful
+	}
+	if completions == *job.Spec.Completions {
+		job.Status.Conditions = append(job.Status.Conditions, newCondition())
+	}
+
+	// no need to update the job if the status hasn't changed since last time
+	if job.Status.Active != active || job.Status.Successful != successful || job.Status.Unsuccessful != unsuccessful {
+		job.Status.Active = active
+		job.Status.Successful = successful
+		job.Status.Unsuccessful = unsuccessful
+
+		if err := jm.updateHandler(&job); err != nil {
+			glog.V(2).Infof("Failed to update job %v, requeuing", job.Name)
+			jm.enqueueController(&job)
+		}
+	}
+	return nil
+}
+
+func newCondition() experimental.JobCondition {
+	return experimental.JobCondition{
+		Type:               experimental.JobComplete,
+		Status:             api.ConditionTrue,
+		LastProbeTime:      util.Now(),
+		LastTransitionTime: util.Now(),
+	}
+}
+
+func getStatus(jobKey string, restartPolicy api.RestartPolicy, pods []api.Pod) (successful, unsuccessful int) {
+	successful = filterPods(pods, api.PodSucceeded)
+	if restartPolicy == api.RestartPolicyNever {
+		unsuccessful = filterPods(pods, api.PodFailed)
+	}
+	return
+}
+
+func (jm *JobManager) manageJob(activePods []*api.Pod, successful, unsuccessful int, job *experimental.Job) int {
+	active := len(activePods)
+	parallelism := *job.Spec.Parallelism
```

Comment:
Dereferencing without a nil check worries me.  Is everyone else satisfied with knowing that if the object is created via conversion from a serialized type and if no one ever clears the field, this works?  It worries me that this code depends on a particular path of building its objects.